chunk/chunk_test.go

```diff
@@ -1,45 +1,111 @@
 package chunk
 
 import (
-	//"bytes"
-	//"io/ioutil"
-	"reflect"
 	"testing"
 	"time"
 
 	"github.com/prometheus/common/model"
 	"github.com/prometheus/prometheus/storage/local/chunk"
-
-	"github.com/weaveworks/common/test"
+	"github.com/stretchr/testify/require"
 )
 
-func TestChunkCodec(t *testing.T) {
+const userID = "userID"
+
+func dummyChunk() Chunk {
+	return dummyChunkFor(model.Metric{
+		model.MetricNameLabel: "foo",
+		"bar":  "baz",
+		"toms": "code",
+	})
+}
+
+func dummyChunkFor(metric model.Metric) Chunk {
 	now := model.Now()
 	cs, _ := chunk.New().Add(model.SamplePair{Timestamp: now, Value: 0})
-	want := Chunk{
-		ID:      "",
-		From:    now.Add(-time.Hour),
-		Through: now,
-		Metric: model.Metric{
-			model.MetricNameLabel: "foo",
-			"bar":  "baz",
-			"toms": "code",
+	chunk := NewChunk(
+		userID,
+		metric.Fingerprint(),
+		metric,
+		cs[0],
+		now.Add(-time.Hour),
+		now,
+	)
+	return chunk
+}
+
+func TestChunkCodec(t *testing.T) {
+	for _, c := range []struct {
+		chunk Chunk
+		err   error
+		f     func(*Chunk, []byte)
+	}{
+		// Basic round trip
+		{chunk: dummyChunk()},
+
+		// Checksum should fail
+		{
+			chunk: dummyChunk(),
+			err:   ErrInvalidChecksum,
+			f:     func(_ *Chunk, buf []byte) { buf[4]++ },
 		},
-		Encoding: chunk.DoubleDelta,
-		Data:     cs[0],
-	}
 
-	r, err := want.reader()
-	if err != nil {
-		t.Fatalf("reader() error: %v", err)
-	}
+		// Checksum should fail
+		{
+			chunk: dummyChunk(),
+			err:   ErrInvalidChecksum,
+			f:     func(c *Chunk, _ []byte) { c.Checksum = 123 },
+		},
+
+		// Metadata test should fail
+		{
+			chunk: dummyChunk(),
+			err:   ErrWrongMetadata,
+			f:     func(c *Chunk, _ []byte) { c.Fingerprint++ },
+		},
+	} {
+		buf, err := c.chunk.encode()
+		require.NoError(t, err)
+
+		have, err := parseExternalKey("", c.chunk.externalKey())
+		require.NoError(t, err)
 
-	have := Chunk{}
-	if err := have.decode(r); err != nil {
-		t.Fatalf("decode() error: %v", err)
+		if c.f != nil {
+			c.f(&have, buf)
+		}
+
+		err = have.decode(buf)
+		require.Equal(t, err, c.err)
+
+		if c.err == nil {
+			require.Equal(t, have, c.chunk)
+		}
 	}
+}
+
+func TestParseExternalKey(t *testing.T) {
+	for _, c := range []struct {
+		key   string
+		chunk Chunk
+		err   error
+	}{
+		{key: "2:1484661279394:1484664879394", chunk: Chunk{
+			UserID:      userID,
+			Fingerprint: model.Fingerprint(2),
+			From:        model.Time(1484661279394),
+			Through:     model.Time(1484664879394),
+		}},
 
-	if !reflect.DeepEqual(want, have) {
-		t.Fatalf("wrong chunks - " + test.Diff(want, have))
+		{key: "1/2:270d8f00:270d8f00:f84c5745", chunk: Chunk{
+			UserID:      "1",
+			Fingerprint: model.Fingerprint(2),
+			From:        model.Time(655200000),
+			Through:     model.Time(655200000),
+			ChecksumSet: true,
+			Checksum:    4165752645,
+		}},
+	} {
+		chunk, err := parseExternalKey(userID, c.key)
+		require.Equal(t, c.err, err)
+		require.Equal(t, c.chunk, chunk)
```

Comment:
Ah, but for legacy chunks it won't - as they don't have the user id prefix.  Although thats not a problem, as we're never writing them out to dynamodb.